academy/ibc-dev/ibcapp_checkers.md

```diff
@@ -0,0 +1,342 @@
+  // validate packet data upon receiving
+  if err := data.ValidateBasic(); err != nil {
+    return packetAck, err
+  }
+
+  score:= types.WinningPlayer {
+    PlayerAddress: data.PlayerAddress,
+    WonCount: data.WonCount,
+    DateAdded: data.DateAdded,
+  }
+
+  leaderboard, found := k.GetLeaderboard(ctx)
+  
+  if !found {
+    panic("Leaderboard not found")
+  }
+
+  found_in_leaderboard := false
+  for i := range leaderboard.Winners {
+      if leaderboard.Winners[i].PlayerAddress == data.PlayerAddress {
+          // found the player, update the fields
+          leaderboard.Winners[i].WonCount = data.WonCount
+          leaderboard.Winners[i].DateAdded = data.DateAdded
+          found_in_leaderboard = true
+          break
+      }
+  }
+
+  // we cannot find the player in the leaderboard
+  if(!found_in_leaderboard) {
+    updated:= append(leaderboard.Winners, &score)
+    leaderboard.Winners = updated
+  } 
+    
+  k.SetLeaderboard(ctx, leaderboard)
+
+  return packetAck, nil
+}
+
+...
+
+```
+
+Above you can see that we differentiate between two cases: either the player is not in the leaderboard yet so we need to append the player and the score or the player can be found in the leaderboard and we need to update the player state. That is it basically. Now the `checkers` blockchain will allow you to send and receive score packets. You can add some additional logic, e.g. check if the packet is valid in the `x/checkers/types/packet_score.go`:
+
+```golang
+...
+
+// ValidateBasic is used for validating the packet
+func (p ScorePacketData) ValidateBasic() error {
+
+  // TODO: Validate the packet data
+
+  // return error if player address is empty
+  if p.PlayerAddress == "" {
+    return errors.New("Player address cannot be empty")
+  }
+
+  return nil
+}
+
+...
+```
+
+## Adjust `leaderboard`
+
+In the `leaderboard` chain, you need first to do the same steps as in the [Migration](add_link) section for the `checkers` chain and include a leaderboard but you will need a slightly different leaderboard structure for the `leaderboard` chain:
+
+```
+message Board {
+  repeated PlayerInfo player = 1; 
+  string GameID = 2; 
+}
+```
+
+This way, you keep also track of the `GameID` and as you can guess, the `leaderboard` chain will maintain different leaderboards with different `GameID`s for different chains.
+
+So you can use the same logic as in the `checkers` chain to send and receive packets. The difference is that for the `leaderboard` chain, the only way to get into one of it's leaderboard is by sending score packets from a game chain.(In `checkers` chain, you will get into the leaderboard structure if you win enough games.)
+
+Anyway we will keep the sending part untouched because we are interested in receiving packets. In the `x/leaderboard/keeper/ibc_top_rank.go` you can find almost the same code as in the `x/checkers/keeper/score.go`:
```

Comment:
I would still let the Leaderboard chain send score packets, because it could be sending them to other leaderboard chains.

Reply:
Yeah that is a use case I did not think of, thanks.

Reply:
My thinking behind only letting checkers send score packets to leaderboard and restricting the leaderboard chain to receiving them, is that we can showcase asymmetric IBC packet flows (Leaderboard throws an error if trying to send or acknowledge a `sendScore` packet